ocaml/libs/xapi-stdext/lib/xapi-stdext-threads/delay_stubs.c

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (C) 2024 Cloud Software Group
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ */
+
+#include <caml/memory.h>
+#include <caml/threads.h>
+#include <caml/custom.h>
+#include <caml/unixsupport.h>
+
+#include <stdbool.h>
+#include <errno.h>
+#include <pthread.h>
+
+typedef struct delay {
+	pthread_mutex_t mtx;
+	pthread_cond_t cond;
+	bool signaled;
+} delay;
+
+// Initialize delay
+// Returns error number or 0 if success
+static int delay_init(delay *d)
+{
+	int err;
+	pthread_condattr_t cond_attr;
+
+	d->signaled = false;
+
+	err = pthread_condattr_init(&cond_attr);
+	if (err)
+		goto err0;
+	err = pthread_condattr_setclock(&cond_attr, CLOCK_MONOTONIC);
+	if (!err)
+		err = pthread_cond_init(&d->cond, &cond_attr);
+	if (err)
+		goto err1;
+	err = pthread_mutex_init(&d->mtx, NULL);
+	if (err)
+		goto err2;
+	pthread_condattr_destroy(&cond_attr);
+	return 0;
+
+err2:
+	pthread_cond_destroy(&d->cond);
+err1:
+	pthread_condattr_destroy(&cond_attr);
+err0:
+	return err;
+}
+
+static void delay_destroy(delay *d)
+{
+	pthread_cond_destroy(&d->cond);
+	pthread_mutex_destroy(&d->mtx);
+}
+
+static void delay_signal(delay *d)
+{
+	// there are quite some chances lock is not held
+	if (pthread_mutex_trylock(&d->mtx) == 0) {
+		d->signaled = true;
+		pthread_cond_signal(&d->cond);
+		pthread_mutex_unlock(&d->mtx);
+		return;
+	}
+
+	// slow way, release engine
+	caml_release_runtime_system();
+	pthread_mutex_lock(&d->mtx);
+	d->signaled = true;
+	pthread_cond_signal(&d->cond);
+	pthread_mutex_unlock(&d->mtx);
```

Comment:
Although unlikely this can fail too, and we should check its return value (which is why I mentioned above that it'd be preferable if we can keep using the OCaml Mutex: error checking is more easily done in OCaml than C: if we want to do error checking correctly in C then it becomes a maze of if/goto to check errors and we lose sight of the actual code we want to run).